p4include/psa.p4

```diff
@@ -0,0 +1,675 @@
+ */
+extern ActionProfile {
+  /// Construct an action profile of 'size' entries
+  ActionProfile(bit<32> size);
+
+  /*
+  @ControlPlaneAPI
+  {
+     entry_handle add_member    (action_ref, action_data);
+     void         delete_member (entry_handle);
+     entry_handle modify_member (entry_handle, action_ref, action_data);
+  }
+  */
+}
+
+/**
+ *  Action selectors are used as table implementation attributes
+ *
+ * Action selectors implement another mechanism to populate table
+ * entries with actions and action data. They are similar to action
+ * profiles, with additional support to define groups of
+ * entries. Action selectors require a hash algorithm to select
+ * members in a group. The only data plane operation required is to
+ * instantiate this extern. When the control plane adds entries
+ * (members) into the extern, they are essentially populating the
+ * corresponding table entries.
+ */
+extern ActionSelector {
+  /// Construct an action selector of 'size' entries
+  /// @param algo hash algorithm to select a member in a group
+  /// @param size number of entries in the action selector
+  /// @param outputWidth size of the key
+  ActionSelector(HashAlgorithm algo, bit<32> size, bit<32> outputWidth);
+
+  /*
+  @ControlPlaneAPI
+  {
+     entry_handle add_member        (action_ref, action_data);
+     void         delete_member     (entry_handle);
+     entry_handle modify_member     (entry_handle, action_ref, action_data);
+     group_handle create_group      ();
+     void         delete_group      (group_handle);
+     void         add_to_group      (group_handle, entry_handle);
+     void         delete_from_group (group_handle, entry_handle);
+  }
+  */
+}
+
+
+
+/* --------------------------------------------------------------------- */
+/// \TODO: is generating a new packet and sending it to the stream or is
+/// it adding a header to the current packet and sending it to the
+/// stream (copying or redirecting).
+extern Digest<T> {
+  Digest(PortId_t receiver); /// define a digest stream to receiver
+  void emit(in T data);      /// emit data into the stream
+
+  /*
+  @ControlPlaneAPI
+  {
+  // TBD
+  // If the type T is a named struct, the name should be used
+  // to generate the control-plane API.
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Programmable blocks
+ *
+ * The following declarations provide a template for the programmable
+ * blocks in the PSA. The P4 programmer is responsible for
+ * implementing controls that match these interfaces and instantiate
+ * them in a package definition.
+ *
+ * The current implementation uses the same user-defined metadata
+ * structure for all the controls.
+ * An alternative design is to split the user-defined metadata into an
+ * input parameter and an output parameter for each block. The
+ * compiler will have to check that the out parameter of a block
+ * matches the in parameter of the subsequent block.
+ *
```

Comment:
The advantage of defining metadata as subsets is that it sets clear bounds on the lifetime of the fields. Good liveness analysis in the compiler can make up for it. The disadvantage, as you say is the need to copy. However, if you have intrinsic metadata directly supported in the hardware, the copies will be unnecessary.  In such cases, the analysis to remove redundant copies is quite trivial. So, I think that in terms of performance the two approaches are equivalent, the analysis are both doable, and it really comes down to a matter of taste! :) And given that we currently already support the flattened view in the compiler, I took the easy path out ...